Dockerfile

```diff
@@ -1,25 +1,22 @@
-FROM ubuntu:14.04
+FROM node:0.10-slim
 ENV PORT 8080
 EXPOSE 8080
-ENV PATH ./node_modules/gulp/bin:$PATH
 ENV NODE_ENV production
 
-# Install apps and dependencies
+# Install app dependencies
 # NOTE: These steps depends highly on service in question
-RUN apt-get update
-RUN apt-get install -y nodejs npm git vim ruby1.9.3 ruby-sass ruby-compass
-RUN ln -s /usr/bin/nodejs /usr/bin/node
 
-# Run app as a custom user `app`
+# Switch to non-root user and create a directory for it
 RUN useradd -m -d /app app
-ADD . /app
-WORKDIR /app
+RUN mkdir -p /app
 RUN chown -R app.app /app
 USER app
-ENV HOME /app
-RUN npm install gulp
-RUN ln -s /app/node_modules/gulp/bin/gulp.js /app/node_modules/gulp/bin/gulp
-RUN npm install
-RUN npm run-script build
-# Default command to run service
-CMD npm start
\ No newline at end of file
+
+# Install and package production deps; see the extensive 
+# file ignore list at .dockerignore
+WORKDIR /app
+COPY . /app/
+RUN npm install --production
```

Comment:
Most of the time npm install takes most of the time in build process. Why not take advance of caching and use two-phased installation:

```
ADD package.json /app/package.json
RUN npm install --production
ADD . /app
```
Or does the usage of COPY avoid re-installation within container? Does it truly work cross-platform?